app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt

```diff
@@ -546,80 +540,82 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
      *
      * @param listRemote function to list remote resources (for instance, all since a certain sync-token)
      */
-    protected open fun syncRemote(listRemote: (MultiResponseCallback) -> Unit) {
-        runBlocking {
-            // download queue
-            val toDownload = LinkedBlockingQueue<HttpUrl>()
-            fun download(url: HttpUrl?) {
-                if (url != null)
-                    toDownload += url
-
-                if (toDownload.size >= MAX_MULTIGET_RESOURCES || url == null) {
-                    while (toDownload.isNotEmpty()) {
-                        val bunch = LinkedList<HttpUrl>()
-                        toDownload.drainTo(bunch, MAX_MULTIGET_RESOURCES)
-                        launch {
-                            downloadRemote(bunch)
-                        }
+    protected open suspend fun syncRemote(listRemote: (MultiResponseCallback) -> Unit) = coroutineScope {    // structured concurrency
+        val syncRemoteScope = this
+
+        // download queue
+        val toDownload = LinkedBlockingQueue<HttpUrl>()
+        fun download(url: HttpUrl?) {
+            if (url != null)
+                toDownload += url
+
+            if (toDownload.size >= MAX_MULTIGET_RESOURCES || url == null) {
+                while (toDownload.isNotEmpty()) {
+                    val bunch = LinkedList<HttpUrl>()
+                    toDownload.drainTo(bunch, MAX_MULTIGET_RESOURCES)
+                    syncRemoteScope.launch {
+                        downloadRemote(bunch)
                     }
                 }
             }
+        }
 
-            coroutineScope {    // structured concurrency: blocks until all inner coroutines are finished
-                listRemote { response, relation ->
-                    // ignore non-members
-                    if (relation != Response.HrefRelation.MEMBER)
-                        return@listRemote
-
-                    // ignore collections
-                    if (response[at.bitfire.dav4jvm.property.webdav.ResourceType::class.java]?.types?.contains(at.bitfire.dav4jvm.property.webdav.ResourceType.COLLECTION) == true)
-                        return@listRemote
-
-                    val name = response.hrefName()
-
-                    if (response.isSuccess()) {
-                        logger.fine("Found remote resource: $name")
-
-                        launch {
-                            val local = localCollection.findByName(name)
-                            SyncException.wrapWithLocalResource(local) {
-                                if (local == null) {
-                                    logger.info("$name has been added remotely, queueing download")
-                                    download(response.href)
+        coroutineScope {    // structured concurrency
+            val listRemoteScope = this
```

Comment:
I don't feel doing this helps with clarity - it actually confused me and had me figure out whether it was necessary. I would use comments for the launch blocks instead.

If you want to keep it like this, I would add a comment here, explaining that this is done for clarity.

Reply:
I agree … I only found it useful for the double-nesting of `coroutineScope` but I guess it's better to remove it again.